### ISSUE TITLE
Signed min/max in solver

### ISSUE THREAD

**Issue by TylerKann** [angr/angr#2252]
I have a simple C function that does the following psuedo code: 
`fun(int a) { `
` return a % 15 }` .
It's pretty obvious the max value returned is 15, while the minimum is 0 (it is returned to rax). I want to see this in angr. I have:
`state = proj.factory.call_state(fun addr)`
`rax1 = claripy.BVS('rax1',64)`
`state.regs.rax = rax1` 
`simgr = proj.factory.simgr(state)`
`simgr.run()`
This gets me to the 1 and only deadened state, which is my return. I want to look at this.
So, I understand I can do `state.se.max(state.regs.rax)` . However, I dont understand this :

 I get 18446744073709551615. Why is it giving me the max unsigned int value? It kind of makes sense, since at the moment there are no constraints. So this brings me to my question. Wouldnt I want to do something like `simgr.deadened.se.max()`? Because I want to look at the registers at the deadened state, aka where simgr is currently; not at state, which is just the start of the function call. Is there a way to do this? I tried doing just `state.step()`, but that wasnt working for me at least. Any thoughts? Thanks !

**Comment by rhelmot** [comment]
The problem here is that the c % operator can produce negative results, and max gives you the max unsigned result. It would be a bit of a refactor to support solving for the max signed result, but it would be possible.

**Comment by rhelmot** [comment]
Also, for the love of god, please use state.solver instead of state.se. state.se has been deprecated for nearly 4 years now...

**Comment by TylerKann** [comment]
@rhelmot Sorry, I did not know that. So in theory just doing state.solver will work? I am slightly confused why I dont need to do some sort of simgr.solver

**Comment by rhelmot** [comment]
yes, state.solver is a straight rename of state.se.

You want simgr.deadended[0].regs.rax etc.

As an additional note, there will be no constraints on the final state. This is because there are no conditional branches in the program you are executing. The fact that the range on rax is restricted is deduced based on the symbolic value in rax, `orig_rax % 15`.

**Comment by TylerKann** [comment]
@rhelmot So is there no way to get a range, if there are no conditional branches? I thought this was the whole point of doing a symbolic bit vectory w/ claripy? 

I really appreciate how helpful you are btw

**Comment by rhelmot** [comment]
No, you can still get the range. As I said, the value in rax is a symbolic expression whose range is implicitly restricted by its construction, so z3 will be able to tell you about it.

If you want to add signed min/max support to angr/claripy, you should start at claripy/backends/backed_z3.py in the `_min` and `_max` functions, adding a `signed` parameter that controls whether the computation uses the ULT/UGT vs the SLT/SGT comparisons. Then, you'll need to add this parameter to every method in the stack between state.solver.min/max and BackendZ3._min/_max.

**Comment by TylerKann** [comment]
@rhelmot sorry I am slightly confused. I thought solver did use z3?

**Comment by rhelmot** [comment]
Solver does use z3. What did I say to indicate that it wouldn't?

**Comment by TylerKann** [comment]
Oh well you said "z3 should be able to tell you about it" I wasnt sure if that meant it was a different feature or I was doing it wrong haha . Signed or unsigned, why would max() produce something more than 15 if it is % 15 ? (positive or negative)

**Comment by rhelmot** [comment]
whether or not the modulo operation is signed or unsigned has no bearing on how the solver interprets your request to find the max. So what will happen in your example is that the possible values that input % 15 can take on is [-14, 14]. However, the largest unsigned number in that range is -1, so that is what max will return.

**Comment by TylerKann** [comment]
So the only way to get a sensible output in this case would be to have an additional signed/unsigned parameter?

**Comment by rhelmot** [comment]
Yes.

**Comment by TylerKann** [comment]
o.o pretty sure I am not up to the challenge of modifying anything but I will definitely look into it.  Thanks for all the help

**Comment by TylerKann** [comment]
just curious, if my code was 
return abs(a % 15) , would that make it much clearer in angr?

**Comment by rhelmot** [comment]
Yes.

**Comment by TylerKann** [comment]
just curious, I changed my C program to the following 
'return abs(a % 15)' so my range is [0,14]. 
I then did 
`state = proj.factory.call_state(addr)'
'rax1 = claripy.BVS('rax1', 64)'
'state.regs.rax = rax1' 
'simgr = proj.factory.simgr(state)'
'simgr.run() % one deadened state'
'simgr.deadened[0].solver.max(simgr.deadened[0].regs.rax)'

It has taken like 45+ min to run the last command (and is still running). Is this normal @rhelmot ? I am assuming I did something wrong bc i dont think the run time should be this long

**Comment by CharlyBVo** [comment]
Hello, 
I began to take a look at this issue. By modifying the backend_z3, I added an option 'signed' using '<' and '>' instead of unsigned operation (hardcoded and not went through all the stack between solver and backend yet). The time required by z3 increases drastically for the simple case % hereunder (I stop it after more than 30 minutes without answers). I manually set a timeout of one second for the solver in the backend and it seems to work by giving the right answer. But it's clearly not the best idea...

Any idea to investigate to reduce z3 solving time ? (I had already tested simplification options of angr for z3).

Small note : In the example above, I try to solve for simgr.deadened[0].regs.**eax**) instead of rax. I think it's something to take into account before trying to solve for a max value --> check where the assembly put the return value to not take into account wrong bits :)
